// lispish.js/src/cons/main.ts
import _cons from "./cons";
import _car from "./car";
import _cdr from "./cdr";
import _print from "./print";
import _equal from "./equal";
import _pair from "./pair";
import _isEmpty from "./isempty";
import _nil from "./nil";
import compose from "../fun/compose";

export const cons = _cons;
export const car = _car;
export const cdr = _cdr;
export const caar = compose(car, car);
export const cadr = compose(car, cdr);
export const cdar = compose(cdr, car);
export const cddr = compose(cdr, cdr);
export const caaar = compose(car, compose(car, car));
export const caadr = compose(car, compose(car, cdr));
export const cadar = compose(car, compose(cdr, car));
export const caddr = compose(car, compose(cdr, cdr));
export const cdaar = compose(cdr, compose(car, car));
export const cdadr = compose(cdr, compose(car, cdr));
export const cddar = compose(cdr, compose(cdr, car));
export const cdddr = compose(cdr, compose(cdr, cdr));
export const caaaar = compose(car, compose(car, compose(car, car)));
export const caaadr = compose(car, compose(car, compose(car, cdr)));
export const caadar = compose(car, compose(car, compose(cdr, car)));
export const caaddr = compose(car, compose(car, compose(cdr, cdr)));
export const cadaar = compose(car, compose(cdr, compose(car, car)));
export const cadadr = compose(car, compose(cdr, compose(car, cdr)));
export const caddar = compose(car, compose(cdr, compose(cdr, car)));
export const cadddr = compose(car, compose(cdr, compose(cdr, cdr)));
export const cdaaar = compose(cdr, compose(car, compose(car, car)));
export const cdaadr = compose(cdr, compose(car, compose(car, cdr)));
export const cdadar = compose(cdr, compose(car, compose(cdr, car)));
export const cdaddr = compose(cdr, compose(car, compose(cdr, cdr)));
export const cddaar = compose(cdr, compose(cdr, compose(car, car)));
export const cddadr = compose(cdr, compose(cdr, compose(car, cdr)));
export const cdddar = compose(cdr, compose(cdr, compose(cdr, car)));
export const cddddr = compose(cdr, compose(cdr, compose(cdr, cdr)));
export const print = _print;
export const equal = _equal;
export const pair = _pair;
export const isEmpty = _isEmpty;
export const nil = _nil;
